src/protocol/context/malicious.rs

```diff
@@ -279,52 +257,151 @@ impl<'a, F: Field> ContextInner<'a, F> {
+        UpgradeContext {
+            upgrade_ctx: self.upgrade_ctx.clone(),
+            inner: self,
+            record_binding: NoRecord,
+        }
+        .upgrade(input)
+        .await
+    }
+
+    async fn upgrade_with_sparse<SS: Substep>(
         &self,
-        record_id: RecordId,
-        x: Replicated<F>,
+        step: &SS,
+        input: Replicated<F>,
         zeros_at: ZeroPositions,
     ) -> Result<MaliciousReplicated<F>, Error> {
-        self.upgrade_one(self.upgrade_ctx.clone(), record_id, x, zeros_at)
-            .await
+        UpgradeContext {
+            upgrade_ctx: self.upgrade_ctx.narrow(step),
+            inner: self,
+            record_binding: NoRecord,
+        }
+        .upgrade_sparse(input, zeros_at)
+        .await
     }
 
-    async fn upgrade_with<SS: Substep>(
+    async fn upgrade_for_record_with<SS: Substep, T, M>(
         &self,
         step: &SS,
         record_id: RecordId,
-        x: Replicated<F>,
+        input: T,
+    ) -> Result<M, Error>
+    where
+        for<'u> UpgradeContext<'u, F, RecordId>: UpgradeToMalicious<T, M>,
+    {
+        UpgradeContext {
+            upgrade_ctx: self.upgrade_ctx.narrow(step),
+            inner: self,
+            record_binding: record_id,
+        }
+        .upgrade(input)
+        .await
+    }
+}
+
+/// Helper to prevent using the record ID multiple times to implement an upgrade.
+///
+/// ```no_run
+/// use raw_ipa::protocol::{context::{NoRecord, UpgradeContext, UpgradeToMalicious}, RecordId};
+/// use raw_ipa::ff::Fp31;
+/// use raw_ipa::secret_sharing::{MaliciousReplicated, Replicated};
+/// let _ = <UpgradeContext<Fp31, NoRecord> as UpgradeToMalicious<Replicated<Fp31>, _>>::upgrade;
+/// let _ = <UpgradeContext<Fp31, RecordId> as UpgradeToMalicious<Replicated<Fp31>, _>>::upgrade;
+/// let _ = <UpgradeContext<Fp31, NoRecord> as UpgradeToMalicious<(Replicated<Fp31>, Replicated<Fp31>), _>>::upgrade;
+/// let _ = <UpgradeContext<Fp31, NoRecord> as UpgradeToMalicious<Vec<Replicated<Fp31>>, _>>::upgrade;
+/// let _ = <UpgradeContext<Fp31, NoRecord> as UpgradeToMalicious<(Vec<Replicated<Fp31>>, Vec<Replicated<Fp31>>), _>>::upgrade;
+/// ```
+///
+/// ```compile_fail
+/// use raw_ipa::protocol::{context::{NoRecord, UpgradeContext, UpgradeToMalicious}, RecordId};
+/// use raw_ipa::ff::Fp31;
+/// use raw_ipa::secret_sharing::{MaliciousReplicated, Replicated};
+/// // This can't be upgraded with a record-bound context because the record ID
+/// // is used internally for vector indexing.
+/// let _ = <UpgradeContext<Fp31, RecordId> as UpgradeToMalicious<Vec<Replicated<Fp31>>, _>>::upgrade;
+/// ```
+pub trait RecordBinding: Copy + Send + Sync {}
+
+#[derive(Clone, Copy)]
+pub struct NoRecord;
+impl RecordBinding for NoRecord {}
+
+impl RecordBinding for RecordId {}
+
+pub struct UpgradeContext<'a, F: Field, B: RecordBinding = NoRecord> {
+    upgrade_ctx: SemiHonestContext<'a, F>,
+    inner: &'a ContextInner<'a, F>,
+    record_binding: B,
+}
```

Comment:
I like this additional structure in general.

I'm not 100% on the idea of binding this way, because the idea that step N has about records might not translate directly to steps that follow on from that.  For instance, our sort effectively takes $N$ input rows and performs $N.2^B$ operations, so each input row effectively maps to more than one operation in a subsequent step.  (The only thing that might save us there is that sorting requires reveals, so we're constantly upgrading and downgrading).

Reply:
For now I only applied the structure to upgrades, which is a somewhat more limited set of use cases than arbitrary protocols. I'm not sure if your comment is referring to upgrades, or to the situation where we try to apply this structure to all contexts? Even for the latter, it's not necessarily a problem. It's fine if the record count varies for different parts of the protocol as long as each of those parts does it's own record_id iteration. The thing that gets messy is if there's an implementation of `foo` that operates on a single record_id (thus takes `record_id` as an argument), that wants to call some other algorithm internally that assigns a different meaning to record IDs. Then we start seeing `a * record_id + b` calculations. I don't think those show up now -- and wouldn't that also be a problem for batching/buffering unless done very carefully?

It does seem like at some point we may want to write some sort of "guide for protocol authors" document that explains some of the non-obvious issues like this one and the need to always use a complete range of [0, total) record IDs.